### ISSUE TITLE
Backreferences appearing before their capture groups incorrectly treated as decimal escapes

### ISSUE THREAD

**Issue by ridiculousfish** [DmitrySoshnikov/regexp-tree#69]
The ES6 spec says that an expression like `\5` is a backreference "only if the integer value of DecimalEscape is <= NcapturingParens", referring to the total number of capturing groups in the entire expression. However regexp-tree only compares to the number of capture groups encountered by the time the backreference is parsed.

See [15.10.2.11_A1_T5  from test262](https://github.com/tc39/test262/blob/da4f4385fdf88ff2c8acf036efaaa62f8cd6bd58/test/built-ins/RegExp/S15.10.2.11_A1_T5.js):

`/\1(A)/.exec("AA");`

This is expected to match the first "A". The backreference is to a capture group that has not captured anything, which this unconditionally suceeds.

Example:

`./bin/regexp-tree -e '/\1(a)/'`

reports:
```
{
  "type": "RegExp",
  "body": {
    "type": "Alternative",
    "expressions": [
      {
        "type": "Char",
        "value": "\\1",
        "kind": "decimal",
        "symbol": "\u0001"
      },
      {
        "type": "Group",
        "capturing": true,
        "number": 1,
        "expression": {
          "type": "Char",
          "value": "a",
          "kind": "simple"
        }
      }
    ]
  },
  "flags": ""
}
```

The first expression should be a back reference.

**Comment by DmitrySoshnikov** [comment]
Yeah, confirm, thanks for catching this; though, it's a rare case, and in fact is a "forwardreference". Need to fix though.

**Comment by MangaD** [comment]
Waiting for the fix. :)

**Comment by datamoon** [comment]
This could be fixed with a two-pass approach:
- a quick pass 1 that only collects group numbers and names, and
- a full pass 2 that can start with knowledge about all available group names and numbers found in pass 1.
